### ISSUE TITLE
Performance of tsserver

### ISSUE THREAD

**Issue by CyberPhoenix90** [Microsoft/TypeScript#25435]
**TypeScript Version: 2.9.2

**Search Terms:** 
Performance, slow, go to definition

**Code**
The codebase we use is closed source and has around 250k lines of typescript spread across many modules some large some small

**Expected behavior:**
Suggestions and go to definition should be fast

**Actual behavior:**
When attempting to get autocomplete or go to definition in visual studio code or webstorm, sometimes it works instantly but more and more we experience huge delays of upwards of one minute for the autocomplete or go to definition to resolve

This could technically be the IDE doing something wrong but since it happens in 2 different IDEs we assume that the delays are coming from tsserver

We use declaration maps and it also looks like go to definition is noticably slower with declaration maps enabled but it's not the full problem

At my workplace we use typescript for all our front end development and it works pretty well for the most part but lately we've been getting crippling slowness. I hope there is something you guys can do about it.
If you need more information let me know I could help debugging what is going on inside tsserver
![image](https://user-images.githubusercontent.com/16382065/42281637-70717d1a-7fa4-11e8-9e43-617fcc342489.png)

here is a snapshot from the process explorer of vscode while the problem is occuring
The system has 4 cores 8 Threads so the CPU usage of tsservice is probably maxed out

**Comment by jason0x43** [comment]
I believe I've been experiencing the same issue. I use TS 2.9.2 with the YouCompleteMe vim plugin. Lately (the last few weeks) I've been getting a lot of timeout errors from the ycmd process YCM uses to talk to tsserver when trying to jump to definitions. The issue seems to be that requests are simply taking a long time (10+ seconds) so YCM gives up. The project I'm working with is much smaller, more like 32k lines of TS.

**Comment by sheetalkamat** [comment]
Please share the tsserver log using 
1. With "typescript.tsserver.log": "verbose" enabled
2. Restart vscode or vscode insiders
3. Reproduce the problem
4. Run ```code --upload-logs``` on the command line (or ```code-insiders --upload-logs``` if using vscode insiders) and follow instructions
5. Share the log file ID

**Comment by CyberPhoenix90** [comment]
Log file ID: 1003061497
If you need any further information let me know

**Comment by sheetalkamat** [comment]
Looking into the log it seems like you have lot of file open request and in all the logs those were the ones that were taking more than half second. None of the operation in tsserver (apart from open) is taking more than half second which makes me think this bug is duplicate of #19458 where in the completion request isnt reaching tsserver while getErr request is running. This needs cancellation from vscode side of "getErr" command before queuing completions. @mjbvz  Seems like this is running into issues of prioritization of getErr(/not cancelling it for completions).

**Comment by CyberPhoenix90** [comment]
I opened files 1 by 1 on purpose to trigger the bug. Each time it took between 3-5 seconds for the editor to resolve the request. The memory of the tsservice process increased by 100 - 150 MB each time I opened a file that wasn't already in the cache once it hits and at a certain point it grinds to a halt when the memory is reaching 1.5 GB. I set my environment nodejs memory cap to 8 GB to see if it helps and I got tsservice to reach 2.5 GB of memory use before it completely stopped responding

It is possible that the issue is duplicate of #19458
Just stating my observations in case those can help

**Comment by sheetalkamat** [comment]
The file open speed depends on how big your project is since it need to read files and bring them in the memory and parse.. Just closing the file does not delete the project from memory and that is intentional since in most cases same project would be re-opened on next file open request. So the orphan projects and scriptinofs are deleted only after open request is completed keeping projects alive in the memory for closed files. In the issue you had reported that goto defs/completions is something that you have issue with and hence mention of #19458 which tracks the issue with completion being not able to reach tsserver because of ongoing getErr request that isnt cancelled.

**Comment by sheetalkamat** [comment]
Moving this to 3.1 as #19458 is also 3.1

**Comment by sheetalkamat** [comment]
Closing this since duplicate is closed as well. Please try with latest typescript and vscode drops. Thanks.
